### PR TITLE
[SPARK-11088] [SQL] Merges partition values using UnsafeProjection

### PR DESCRIPTION
**Pull request by liancheng**
`DataSourceStrategy.mergeWithPartitionValues` is essentially a projection implemented in a quite inefficient way. This PR optimizes this method with `UnsafeProjection` to avoid unnecessary boxing costs.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 1]
```diff
@@ -178,52 +179,26 @@ private[sql] object DataSourceStrategy extends Strategy with Logging {
...
-        iterator.map { dataRow =>
-          var i = 0
-          while (i < mutableRow.numFields) {
-            mergers(i)(mutableRow, dataRow, i)
-            i += 1
-          }
-          mutableRow.asInstanceOf[InternalRow]
-        }
+        val projection = UnsafeProjection.create(requiredColumns, dataColumns ++ partitionColumns)
+        iterator.map(dataRow => projection(new JoinedRow(dataRow, partitionValues)))
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: Do we have to allocate a new JoinedRow each time?

**Comment by liancheng** [comment]
Micro-benchmark result with TPC-DS (scale-factor 15) `store_sales` table shows a ~12% performance gain.

Before:

- Round 0: 8133 ms
- Round 1: 7799 ms
- Round 2: 8010 ms
- Round 3: 8009 ms
- Round 4: 8223 ms
- Average: 8034.8 ms

After:

- Round 0: 7401 ms
- Round 1: 6897 ms
- Round 2: 6873 ms
- Round 3: 6935 ms
- Round 4: 7056 ms
- Average: 7032.4 ms

Benchmark code (where `ss_sold_date_sk` is an `INT` partitioning column and `ss_sold_time_sk` is an `INT` data column):

```scala
import com.google.common.base.Stopwatch

def benchmark(runs: Int, warmupRuns: Int = 0)(f: => Unit) {
  val stopwatch = new Stopwatch()

  (0 until warmupRuns).foreach { i =>
    f
  }

  def run(i: Int) = {
    stopwatch.reset()
    stopwatch.start()
    f
    stopwatch.stop()
    val elapsed = stopwatch.elapsedMillis()
    println(s"Round $i: $elapsed ms")
    elapsed
  }

  val total = (0 until runs).map(i => run(i)).sum.toDouble
  println(s"Average: ${total / runs} ms")
}

val path = "file:///Users/lian/tpcds/sf15/store_sales"

benchmark(5, 5) {
  val df = sqlContext.read.parquet(path).selectExpr("ss_sold_time_sk", "ss_sold_date_sk")
  df.queryExecution.toRdd.foreach(row => ())
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala (hunk 1 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: That's a good point, didn't realize `JoinedRow` is mutable. Thanks.

**Comment by marmbrus** [comment]
Can we also remove an extra `ConvertToUnsafe` here?  Specifically, is the parquet table scan still claiming to produce safe rows when its really producing unsafe ones now?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala (hunk 1 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: Updated, although reusing `JoinedRow` doesn't bring noticeable speedup in my micro-benchmark.

**Comment by liancheng** [comment]
@marmbrus Addressed the `ConvertToUnsafe` issue in PR #9125, since this is mostly a separate issue.

**Comment by marmbrus** [comment]
LGTM, merging to master.

**Pull request closed without merging**